MIDI-149: Nano GPT training pipeline

WojciechMat opened the pull request:
### nanoGPT
I remade configuration logic to use hydra.

To run the training:

```sh
python train.py
```

To run the output monitoring dashboard:

```sh
PYTHONPATH=. streamlit run --server.port 4002 dashboard/main.py
```

To use AwesomeTokensDataset, first run
```sh
python -m scripts.train_awesome_tokenizer  # creates a pre-trained tokenizer json in pretrained/awesome_tokenizers
```
You'll also have to setup an `.env` file for wandb and huggingface, follow the `.env.example` pattern with your personal tokens.
### Augmentation
I use `pitch_shift` and `change_speed`. I apply it to the original midi dataset for each record with `augmentation_probability=0.2`, `augmentation_repetitions=5`, sequentially (pitch_shift, then change_speed). It should result in **~4 times larger** dataset.

E.g. number of tokens in basic AwesomeTokensDataset:
| Augmentation | Train      | Test   | Validation |
|------|------------|--------|------------|
| on | 10,647,040 | 241,152| 288,256    |
| off | 2,614,272 | 241,152| 288,256    |

Augmenting maestro dataset takes ~8 minute, so augmenting giant_midi + maestro should take ~80 minutes...
If you think of any optimization methods, do let me know.

WojciechMat committed:
exact copy of nanoGPT train and model, not working

WojciechMat committed:
minimal changes to make it work

WojciechMat committed:
update .gitignore

WojciechMat committed:
add hydra to requirements

WojciechMat committed:
requirements.txt

WojciechMat committed:
fix wandb

WojciechMat committed:
logging bug

WojciechMat committed:
merge

WojciechMat committed:
fix config bug, add tqdm

WojciechMat committed:
tokens per second

WojciechMat committed:
spaces

WojciechMat committed:
cuda:0 -> cuda in config

WojciechMat committed:
update checkpoints directory:

WojciechMat committed:
encoded datasets

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
fix path

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
initial version of model review

WojciechMat committed:
AwesomeTokensDataset

WojciechMat committed:
main dashboard

WojciechMat committed:
configs

WojciechMat committed:
wrong config key name

WojciechMat committed:
fix config

WojciechMat committed:
remove tqdm from train.py

WojciechMat committed:
rename the configs, fix f-string, use only maestro when training awesome tokenizer

roszcz commented on train.py:
```diff
@@ -1,18 +1,389 @@
+    total_tokens = dataset_config.sequence_length * dataset["train"].num_rows
+    print(f"tokens in a training dataset: {total_tokens}")
+    if cfg.architecture != "custom":
+        architecture_config_path = f"configs/architectures/{cfg.architecture}.yaml"
+        architecture_config_path = to_absolute_path(architecture_config_path)
+        with open(architecture_config_path, "r") as file:
+            architecture_config = yaml.safe_load(file)
+
+        # override model config
+        cfg.model.n_layer = architecture_config["n_layer"]
+        cfg.model.n_head = architecture_config["n_head"]
+        cfg.model.n_embd = architecture_config["n_embd"]
```
Looks like this is not using hydra correctly - you should be able to load any architecture config without manually reading it as yaml, but by defining a proper relation in hydra: https://hydra.cc/docs/tutorials/basic/your_first_app/composition/

WojciechMat committed:
fix dashboard

WojciechMat committed:
better dashboard: untokenization with context, list of devices

WojciechMat committed:
change tokenizer path

WojciechMat committed:
make pianorolls look better - v1

WojciechMat committed:
merge

WojciechMat committed:
.env.example, remove tqdm, initial analyze_datasets script for help

WojciechMat committed:
fix hydra, play full model output (to be changed)

WojciechMat committed:
better scripts, remove yaml from train.py

WojciechMat committed:
min_time_unit 0.001 -> 0.01, bigger segmentation steps

WojciechMat committed:
update configs, remove fixing tokens

WojciechMat committed:
shift generated notes, fix warnings by adding .copy() to iloc

WojciechMat committed:
download button

WojciechMat committed:
tmp directory

WojciechMat committed:
download midi with context button

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
secondary piece in pianoroll, move gpt2 logic to a module

WojciechMat committed:
config structuring

WojciechMat committed:
prepare init_from for finetuning

WojciechMat committed:
pretrained directory

WojciechMat committed:
augmentation, midi with context fixed

WojciechMat committed:
augmentation

WojciechMat committed:
cache all datasets, augmentation

WojciechMat committed:
fix mididatasert, add configs for colossal dataset

WojciechMat committed:
training commands and configs

WojciechMat committed:
better masks

WojciechMat committed:
fix config

WojciechMat committed:
load dotenv

WojciechMat committed:
total_tokens metric

WojciechMat committed:
add new datasets to caching script

WojciechMat committed:
evaluation dashboard

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
requirements, update training_plan

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
use cfg instead of train_config

WojciechMat committed:
AwesomeTokenizer -> AwesomeMidiTokenizer

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
awesome tokenizer hardcoded pretrained

WojciechMat committed:
misspell

WojciechMat committed:
dotenv in script

WojciechMat committed:
pretrained tokenizer to absolute path

WojciechMat committed:
save cpus

WojciechMat committed:
fix tokens per second

WojciechMat committed:
fix awesome dashboard

WojciechMat committed:
more frequent logging, tensors to numpy, remove ifs from training loop

WojciechMat committed:
bug...

WojciechMat committed:
tensors should first be moved to cpu, than numpy

WojciechMat committed:
move output tensor to cpu at once

WojciechMat committed:
get model parameters before packing it to ddp

WojciechMat committed:
download model button, milion parameters in midi name

WojciechMat committed:
bad ideal

roszcz committed:
alternative approach to token counting

- counting tokens within microbatches (is there a reson not to?)
- show vest loss per checkpoint on the dashboard
- more aggresive cpu allocation for augmentation

WojciechMat committed:
dataset configs in hydra

WojciechMat committed:
only load_tokenzier methods in datasets

WojciechMat committed:
awesome tokenziers non-heuristically

WojciechMat committed:
dataset selection, no augmentation on dashboard, no extra datasets on dashboard

WojciechMat committed:
misspell

WojciechMat committed:
another misspell...

WojciechMat committed:
download button for source file

WojciechMat committed:
select piece

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
add index to a filenames

WojciechMat committed:
better st.form

WojciechMat committed:
generate midi from file script

WojciechMat committed:
generation script

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt

WojciechMat committed:
updates

WojciechMat committed:
new midi files

WojciechMat committed:
change example midi files

WojciechMat committed:
NoLoss -> ExponentialTime

WojciechMat committed:
update requirements

WojciechMat committed:
update requirements

WojciechMat committed:
add docstrings

WojciechMat committed:
backward compatibility

WojciechMat committed:
add README

WojciechMat committed:
fix some issues

WojciechMat committed:
fix hf caching, change configs

WojciechMat committed:
new generated midi

WojciechMat committed:
Merge branch 'MIDI-149/nano-gpt' of github.com:Nospoko/midi-transformer into MIDI-149/nano-gpt
